From d3883b1f653de9fd88fc457f7275e287cc1e985f Mon Sep 17 00:00:00 2001
From: Yu Qiaowei <cerf.yu@rock-chips.com>
Date: Wed, 8 Apr 2026 20:22:45 +0800
Subject: [PATCH 19/39] video: rockchip: rga3: support hardware batching

Users can enable this by setting RGA_REQUEST_FLAGS_EXEC_SEQUENTIAL. The
current request is batched into hardware commands, running in
master_mode on RGA2/RGA3.

Change-Id: I3cfb00b8e8e0c5fb7c36eb11425ac27ae42ad7e7
Signed-off-by: Yu Qiaowei <cerf.yu@rock-chips.com>
(cherry picked from commit 02e0554b1e662c2da6e44fd00376892cbde8dc59)
---
 drivers/video/rockchip/rga3/include/rga.h     |   4 +
 .../rockchip/rga3/include/rga2_reg_info.h     |  18 +-
 .../rockchip/rga3/include/rga3_reg_info.h     |   9 +
 drivers/video/rockchip/rga3/include/rga_drv.h |  23 +-
 .../video/rockchip/rga3/include/rga_iommu.h   |   4 +-
 drivers/video/rockchip/rga3/include/rga_job.h |   2 +-
 drivers/video/rockchip/rga3/rga2_reg_info.c   | 339 ++++++++++++------
 drivers/video/rockchip/rga3/rga3_reg_info.c   | 179 +++++----
 drivers/video/rockchip/rga3/rga_debugger.c    |  32 +-
 drivers/video/rockchip/rga3/rga_iommu.c       |  44 +--
 drivers/video/rockchip/rga3/rga_job.c         | 138 ++++---
 drivers/video/rockchip/rga3/rga_mm.c          | 251 +++++++------
 drivers/video/rockchip/rga3/rga_policy.c      |  42 ++-
 13 files changed, 704 insertions(+), 381 deletions(-)

diff --git a/drivers/video/rockchip/rga3/include/rga.h b/drivers/video/rockchip/rga3/include/rga.h
index 671867b22371..3d185ac7ab28 100644
--- a/drivers/video/rockchip/rga3/include/rga.h
+++ b/drivers/video/rockchip/rga3/include/rga.h
@@ -138,6 +138,7 @@ enum {
 	RGA_10BIT_INCOMPACT		= 0x1,
 };
 
+/* MPI context flags(request flags) */
 enum {
 	RGA_CONTEXT_NONE		= 0x0,
 	RGA_CONTEXT_SRC_FIX_ENABLE	= 0x1 << 0,
@@ -154,6 +155,9 @@ enum {
 					  RGA_CONTEXT_DST_CACHE_INFO,
 };
 
+/* request flags */
+#define RGA_REQUEST_FLAGS_EXEC_SEQUENTIAL	BIT(6)
+
 /* RGA feature */
 enum {
 	RGA_COLOR_FILL			= 0x1 << 0,
diff --git a/drivers/video/rockchip/rga3/include/rga2_reg_info.h b/drivers/video/rockchip/rga3/include/rga2_reg_info.h
index 187460833cf5..5f4238b84eb6 100644
--- a/drivers/video/rockchip/rga3/include/rga2_reg_info.h
+++ b/drivers/video/rockchip/rga3/include/rga2_reg_info.h
@@ -127,19 +127,31 @@
 #define s_RGA2_SYS_CTRL_HOLD_MODE_EN(x)			((x & 0x1) << 9)
 #define s_RGA2_SYS_CTRL_CMD_MODE(x)			((x & 0x1) << 1)
 
+#define v_RGA2_SYS_CTRL_CMD_MODE(x)			(((x) >> 1) & 0x1)
+
 /* RGA_CMD_CTRL */
+#define m_RGA2_MAX_RD_OUTSTANDING			(0x3f << 14)
 #define m_RGA2_CMD_CTRL_INCR_NUM			(0x3ff << 3)
 #define m_RGA2_CMD_CTRL_STOP				(0x1 << 2)
 #define m_RGA2_CMD_CTRL_INCR_VALID_P			(0x1 << 1)
 #define m_RGA2_CMD_CTRL_CMD_LINE_ST_P			(0x1 << 0)
 
-#define s_RGA2_CMD_CTRL_INCR_NUM(x)			((x & 0x3ff) << 3)
+#define s_RGA2_MAX_RD_OUTSTANDING(x)			(((x) & 0x3f) << 14)
+#define s_RGA2_CMD_CTRL_INCR_NUM(x)			(((x) & 0x3ff) << 3)
+#define s_RGA2_CMD_CTRL_STOP(x)				(((x) & 0x1) << 2)
+
+#define v_RGA2_MAX_RD_OUTSTANDING(x)			(((x) >> 14) & 0x3f)
+#define v_RGA2_CMD_CTRL_INCR_NUM(x)			(((x) >> 3) & 0x3ff)
+#define v_RGA2_CMD_CTRL_STOP(x)				(((x) >> 2) & 0x1)
 
 /* RGA_STATUS1 */
-#define m_RGA2_STATUS1_SW_CMD_TOTAL_NUM			(0xfff << 8)
+#define m_RGA2_STATUS1_SW_CMD_TOTAL_NUM			(0xfff << 20)
 #define m_RGA2_STATUS1_SW_CMD_CUR_NUM			(0xfff << 8)
 #define m_RGA2_STATUS1_SW_RGA_STA			(0x1 << 0)
 
+#define v_RGA2_STATUS1_SW_CMD_TOTAL_NUM(x)		(((x) >> 20) & 0xfff)
+#define v_RGA2_STATUS1_SW_CMD_CUR_NUM(x)		(((x) >> 8) & 0xfff)
+
 /*RGA_INT*/
 #define m_RGA2_INT_FBCIN_DEC_ERROR_CLEAR		(1 << 24)
 #define m_RGA2_INT_FBCIN_DEC_ERROR_EN			(1 << 23)
@@ -530,6 +542,8 @@
 #define RGA2_IOMMU_PREFETCH_THRESHOLD_MIN		(0x1U << RGA2_IOMMU_PREFETCH_SHIFT)
 #define RGA2_IOMMU_PREFETCH_THRESHOLD_MAX		(0xffffU << RGA2_IOMMU_PREFETCH_SHIFT)
 
+#define RGA2_MAX_RD_OUTSTANDING				0x3f
+
 union rga2_color_ctrl {
 	uint32_t value;
 	struct {
diff --git a/drivers/video/rockchip/rga3/include/rga3_reg_info.h b/drivers/video/rockchip/rga3/include/rga3_reg_info.h
index 50629a3fd994..0c7aabe62392 100644
--- a/drivers/video/rockchip/rga3/include/rga3_reg_info.h
+++ b/drivers/video/rockchip/rga3/include/rga3_reg_info.h
@@ -81,6 +81,8 @@
 #define s_RGA3_SYS_CTRL_ACLK_SRESET(x)				((x & 0x1) << 3)
 #define s_RGA3_SYS_CTRL_CMD_MODE(x)				((x & 0x1) << 1)
 
+#define v_RGA3_SYS_CTRL_CMD_MODE(x)				(((x) >> 1) & 0x1)
+
 /* TODO: RGA3_INT_EN/RGA3_INT_RAW/RGA3_INT_MSK/RGA3_INT_CLR */
 #define m_RGA3_INT_WIN1_VOR_FIFO_REN_ERR			(0x1 << 29)
 #define m_RGA3_INT_WIN1_VOR_FIFO_WEN_ERR			(0x1 << 28)
@@ -143,6 +145,11 @@
 #define m_RGA3_CMD_CTRL_CMD_INCR_VALID_P			(0x1 << 1)
 #define m_RGA3_CMD_CTRL_CMD_LINE_ST_P				(0x1 << 0)
 
+#define s_RGA3_CMD_CTRL_CMD_INCR_NUM(x)				(((x) & 0x3ff) << 3)
+#define s_RGA3_CMD_CTRL_CMD_STOP_MODE(x)			(((x) & 0x1) << 2)
+
+#define v_RGA3_CMD_CTRL_CMD_INCR_NUM(x)				(((x) >> 3) & 0x3ff)
+
 /* RGA3_RO_SRST */
 #define m_RGA3_RO_SRST_RO_RST_DONE				(0x3f << 0)
 
@@ -150,6 +157,8 @@
 #define m_RGA3_CMD_STATE_CMD_CNT_CUR				(0xfff << 16)
 #define m_RGA3_CMD_STATE_CMD_WORKING				(0x1 << 0)
 
+#define v_RGA3_CMD_STATE_CMD_CNT_CUR(x)				(((x) >> 16) & 0xfff)
+
 /* RGA3_WIN0_RD_CTRL */
 #define m_RGA3_WIN0_RD_CTRL_SW_WIN0_ENABLE			(0x1 << 0)
 #define m_RGA3_WIN0_RD_CTRL_SW_WIN0_RD_MODE			(0x3 << 1)
diff --git a/drivers/video/rockchip/rga3/include/rga_drv.h b/drivers/video/rockchip/rga3/include/rga_drv.h
index c5ec6f965638..d2edb30b83d9 100644
--- a/drivers/video/rockchip/rga3/include/rga_drv.h
+++ b/drivers/video/rockchip/rga3/include/rga_drv.h
@@ -279,6 +279,14 @@ struct rga_job_buffer {
 	int page_count;
 };
 
+struct rga_job_task_buffers {
+	struct rga_job_buffer src_buffer;
+	struct rga_job_buffer src1_buffer;
+	struct rga_job_buffer dst_buffer;
+	/* used by rga2 */
+	struct rga_job_buffer els_buffer;
+};
+
 struct rga_job_timestamp {
 	ktime_t init;
 	ktime_t insert;
@@ -296,17 +304,16 @@ struct rga_job {
 	struct rga_scheduler_t *scheduler;
 	struct rga_session *session;
 
-	struct rga_req rga_command_base;
-	struct rga_dma_buffer *cmd_buf;
 	struct rga_full_csc full_csc;
 	struct rga_csc_clip full_csc_clip;
 	struct rga_pre_intr_info pre_intr_info;
 
-	struct rga_job_buffer src_buffer;
-	struct rga_job_buffer src1_buffer;
-	struct rga_job_buffer dst_buffer;
-	/* used by rga2 */
-	struct rga_job_buffer els_buffer;
+	/* multi-task management */
+	struct rga_dma_buffer *cmd_buf;
+	struct rga_req *task_list;
+	struct rga_job_task_buffers *task_buffers;
+	size_t task_count;
+	size_t finished_count;
 
 	/* for rga2 virtual_address */
 	struct mm_struct *mm;
@@ -320,7 +327,6 @@ struct rga_job {
 	int core;
 	int ret;
 	pid_t pid;
-	bool use_batch_mode;
 
 	struct kref refcount;
 	unsigned long state;
@@ -383,7 +389,6 @@ struct rga_request {
 	uint32_t finished_task_count;
 	uint32_t failed_task_count;
 
-	bool use_batch_mode;
 	bool is_running;
 	bool is_done;
 	int ret;
diff --git a/drivers/video/rockchip/rga3/include/rga_iommu.h b/drivers/video/rockchip/rga3/include/rga_iommu.h
index b80a1f48bb25..79f533c20ab5 100644
--- a/drivers/video/rockchip/rga3/include/rga_iommu.h
+++ b/drivers/video/rockchip/rga3/include/rga_iommu.h
@@ -61,7 +61,9 @@ struct rga_mmu_base {
 };
 
 int rga_user_memory_check(struct page **pages, u32 w, u32 h, u32 format, int flag);
-int rga_set_mmu_base(struct rga_job *job, struct rga2_req *req);
+int rga_set_mmu_base(struct rga_job *job,
+		     struct rga_job_task_buffers *task_buffers,
+		     struct rga2_req *req);
 unsigned int *rga_mmu_buf_get(struct rga_mmu_base *mmu_base, uint32_t size);
 
 struct rga_mmu_base *rga_mmu_base_init(size_t size);
diff --git a/drivers/video/rockchip/rga3/include/rga_job.h b/drivers/video/rockchip/rga3/include/rga_job.h
index e63659bb0e8e..ac206c2cb377 100644
--- a/drivers/video/rockchip/rga3/include/rga_job.h
+++ b/drivers/video/rockchip/rga3/include/rga_job.h
@@ -25,7 +25,7 @@ enum job_flags {
 void rga_job_scheduler_dump_info(struct rga_scheduler_t *scheduler);
 void rga_job_next(struct rga_scheduler_t *scheduler);
 struct rga_job *rga_job_done(struct rga_scheduler_t *scheduler);
-int rga_job_commit(struct rga_req *rga_command_base, struct rga_request *request);
+int rga_job_commit(struct rga_req *task_list, size_t task_count, struct rga_request *request);
 int rga_job_mpi_commit(struct rga_req *rga_command_base, struct rga_request *request);
 
 int rga_job_assign(struct rga_job *job);
diff --git a/drivers/video/rockchip/rga3/rga2_reg_info.c b/drivers/video/rockchip/rga3/rga2_reg_info.c
index 5e4c69a67357..501c31a7608d 100644
--- a/drivers/video/rockchip/rga3/rga2_reg_info.c
+++ b/drivers/video/rockchip/rga3/rga2_reg_info.c
@@ -2813,65 +2813,108 @@ static int rga2_align_check(struct rga_job *job, struct rga2_req *req)
 	return 0;
 }
 
-static void print_debug_info(struct rga_job *job, struct rga2_req *req)
+static void print_debug_info(struct rga_job *job, struct rga2_req *req, int task_index)
 {
-	rga_job_log(job, "render_mode:%s,bitblit_mode=%d,rotate_mode:%s\n",
-		rga_get_render_mode_str(req->render_mode), req->bitblt_mode,
+	rga_job_log(job, "task[%d]: render_mode:%s,bitblit_mode=%d,rotate_mode:%s\n",
+		task_index, rga_get_render_mode_str(req->render_mode), req->bitblt_mode,
 		rga_get_rotate_mode_str(req->rotate_mode));
 
-	rga_job_log(job, "src: y=%lx uv=%lx v=%lx aw=%d ah=%d vw=%d vh=%d\n",
-		 (unsigned long)req->src.yrgb_addr,
-		 (unsigned long)req->src.uv_addr,
-		 (unsigned long)req->src.v_addr,
-		 req->src.act_w, req->src.act_h,
-		 req->src.vir_w, req->src.vir_h);
-	rga_job_log(job, "src: xoff=%d yoff=%d format=%s\n",
+	rga_job_log(job, "task[%d]: src: y=%lx uv=%lx v=%lx aw=%d ah=%d vw=%d vh=%d\n",
+		task_index,
+		(unsigned long)req->src.yrgb_addr,
+		(unsigned long)req->src.uv_addr,
+		(unsigned long)req->src.v_addr,
+		req->src.act_w, req->src.act_h,
+		req->src.vir_w, req->src.vir_h);
+	rga_job_log(job, "task[%d]: src: xoff=%d yoff=%d format=%s\n",
+		task_index,
 		req->src.x_offset, req->src.y_offset,
-		 rga_get_format_name(req->src.format));
+		rga_get_format_name(req->src.format));
 
 	if (req->src1.yrgb_addr != 0 || req->src1.uv_addr != 0
 		|| req->src1.v_addr != 0) {
-		rga_job_log(job, "src1: y=%lx uv=%lx v=%lx aw=%d ah=%d vw=%d vh=%d\n",
-			 (unsigned long)req->src1.yrgb_addr,
-			 (unsigned long)req->src1.uv_addr,
-			 (unsigned long)req->src1.v_addr,
-			 req->src1.act_w, req->src1.act_h,
-			 req->src1.vir_w, req->src1.vir_h);
-		rga_job_log(job, "src1: xoff=%d yoff=%d format=%s\n",
-			req->src1.x_offset, req->src1.y_offset,
-			 rga_get_format_name(req->src1.format));
+		rga_job_log(job, "task[%d]: src1: y=%lx uv=%lx v=%lx aw=%d ah=%d vw=%d vh=%d\n",
+			task_index,
+			(unsigned long)req->src1.yrgb_addr,
+			(unsigned long)req->src1.uv_addr,
+			(unsigned long)req->src1.v_addr,
+			req->src1.act_w, req->src1.act_h,
+			req->src1.vir_w, req->src1.vir_h);
+		rga_job_log(job, "task[%d]: src1: xoff=%d yoff=%d format=%s\n",
+			task_index, req->src1.x_offset, req->src1.y_offset,
+			rga_get_format_name(req->src1.format));
 	}
 
-	rga_job_log(job, "dst: y=%lx uv=%lx v=%lx aw=%d ah=%d vw=%d vh=%d\n",
-		 (unsigned long)req->dst.yrgb_addr,
-		 (unsigned long)req->dst.uv_addr,
-		 (unsigned long)req->dst.v_addr,
-		 req->dst.act_w, req->dst.act_h,
-		 req->dst.vir_w, req->dst.vir_h);
-	rga_job_log(job, "dst: xoff=%d yoff=%d format=%s\n",
+	rga_job_log(job, "task[%d]: dst: y=%lx uv=%lx v=%lx aw=%d ah=%d vw=%d vh=%d\n",
+		task_index, (unsigned long)req->dst.yrgb_addr,
+		(unsigned long)req->dst.uv_addr,
+		(unsigned long)req->dst.v_addr,
+		req->dst.act_w, req->dst.act_h,
+		req->dst.vir_w, req->dst.vir_h);
+	rga_job_log(job, "task[%d]: dst: xoff=%d yoff=%d format=%s\n",
+		task_index,
 		req->dst.x_offset, req->dst.y_offset,
-		 rga_get_format_name(req->dst.format));
+		rga_get_format_name(req->dst.format));
 
-	rga_job_log(job, "mmu: src=%.2x src1=%.2x dst=%.2x els=%.2x\n",
+	rga_job_log(job, "task[%d]: mmu: src=%.2x src1=%.2x dst=%.2x els=%.2x\n",
+		task_index,
 		req->mmu_info.src0_mmu_flag, req->mmu_info.src1_mmu_flag,
 		req->mmu_info.dst_mmu_flag, req->mmu_info.els_mmu_flag);
-	rga_job_log(job, "alpha: flag %x mode=%s\n",
+	rga_job_log(job, "task[%d]: alpha: flag %x mode=%s\n",
+		task_index,
 		req->alpha_rop_flag, rga_get_blend_mode_str(req->alpha_config.mode));
-	rga_job_log(job, "alpha: pre_multi=[%d,%d] pixl=[%d,%d] glb=[%d,%d]\n",
+	rga_job_log(job, "task[%d]: alpha: pre_multi=[%d,%d] pixl=[%d,%d] glb=[%d,%d]\n",
+		task_index,
 		req->alpha_config.fg_pre_multiplied, req->alpha_config.bg_pre_multiplied,
 		req->alpha_config.fg_pixel_alpha_en, req->alpha_config.bg_pixel_alpha_en,
 		req->alpha_config.fg_global_alpha_en, req->alpha_config.bg_global_alpha_en);
-	rga_job_log(job, "alpha: fg_global_alpha=%x bg_global_alpha=%x\n",
+	rga_job_log(job, "task[%d]: alpha: fg_global_alpha=%x bg_global_alpha=%x\n",
+		task_index,
 		req->alpha_config.fg_global_alpha_value, req->alpha_config.bg_global_alpha_value);
-	rga_job_log(job, "yuv2rgb mode is %x\n", req->yuv2rgb_mode);
+	rga_job_log(job, "task[%d]: yuv2rgb mode is %x\n", task_index, req->yuv2rgb_mode);
 }
 
-static int rga2_init_reg(struct rga_job *job)
+static int rga2_init_sys_reg(struct rga_job *job)
+{
+	int i;
+	struct rga_req *user_req = NULL;
+
+	/* set default full_csc_clip values */
+	job->full_csc_clip.y.max = 0xff;
+	job->full_csc_clip.y.min = 0x0;
+	job->full_csc_clip.uv.max = 0xff;
+	job->full_csc_clip.uv.min = 0x0;
+
+	for (i = 0; i < job->task_count; i++) {
+		user_req = &job->task_list[i];
+
+		if (user_req->full_csc.flag & 0x1) {
+			if (user_req->feature.full_csc_clip_en) {
+				memcpy(&job->full_csc_clip, &user_req->full_csc_clip,
+				       sizeof(job->full_csc_clip));
+			}
+			memcpy(&job->full_csc, &user_req->full_csc, sizeof(job->full_csc));
+
+			/* only need to set once */
+			break;
+		}
+	}
+
+	memcpy(&job->pre_intr_info, &job->task_list[0].pre_intr_info,
+	       sizeof(job->pre_intr_info));
+
+	return 0;
+}
+
+static int rga2_init_cmd_reg(struct rga_job *job)
 {
-	struct rga2_req req;
 	int ret = 0;
+	int i;
+	uint32_t *cmd_base = NULL;
+	struct rga_req *user_req = NULL;
+	struct rga_job_task_buffers *task_buffers = NULL;
+	struct rga2_req req;
 	struct rga_scheduler_t *scheduler = NULL;
-	ktime_t timestamp = ktime_get();
 
 	scheduler = job->scheduler;
 	if (unlikely(scheduler == NULL)) {
@@ -2879,78 +2922,80 @@ static int rga2_init_reg(struct rga_job *job)
 		return -EINVAL;
 	}
 
-	memset(&req, 0x0, sizeof(req));
+	for (i = 0; i < job->task_count; i++) {
+		user_req = &job->task_list[i];
+		cmd_base = (uint32_t *)job->cmd_buf->vaddr + i * scheduler->data->cmd_reg_size;
+		task_buffers = &job->task_buffers[i];
 
-	rga_cmd_to_rga2_cmd(scheduler, &job->rga_command_base, &req);
-	if (req.full_csc_en) {
-		memcpy(&job->full_csc, &job->rga_command_base.full_csc, sizeof(job->full_csc));
-		if (job->rga_command_base.feature.full_csc_clip_en) {
-			memcpy(&job->full_csc_clip, &job->rga_command_base.full_csc_clip,
-			       sizeof(job->full_csc_clip));
-		} else {
-			job->full_csc_clip.y.max = 0xff;
-			job->full_csc_clip.y.min = 0x0;
-			job->full_csc_clip.uv.max = 0xff;
-			job->full_csc_clip.uv.min = 0x0;
-		}
+		memset(&req, 0x0, sizeof(req));
 
-	} else {
-		job->full_csc_clip.y.max = 0xff;
-		job->full_csc_clip.y.min = 0x0;
-		job->full_csc_clip.uv.max = 0xff;
-		job->full_csc_clip.uv.min = 0x0;
-	}
-	memcpy(&job->pre_intr_info, &job->rga_command_base.pre_intr_info,
-	       sizeof(job->pre_intr_info));
+		rga_cmd_to_rga2_cmd(scheduler, user_req, &req);
 
-	/* for debug */
-	if (DEBUGGER_EN(MSG))
-		print_debug_info(job, &req);
+		/* for debug */
+		if (DEBUGGER_EN(MSG))
+			print_debug_info(job, &req, i);
 
-	/* check value if legal */
-	ret = rga2_check_param(job, scheduler->data, &req);
-	if (ret == -EINVAL) {
-		rga_job_err(job, "req argument is inval\n");
-		return ret;
-	}
+		/* check value if legal */
+		ret = rga2_check_param(job, scheduler->data, &req);
+		if (ret == -EINVAL) {
+			rga_job_err(job, "req argument is invalid\n");
+			return ret;
+		}
 
-	rga2_align_check(job, &req);
-
-	/* RGA2 mmu set */
-	if ((req.mmu_info.src0_mmu_flag & 1) || (req.mmu_info.src1_mmu_flag & 1) ||
-	    (req.mmu_info.dst_mmu_flag & 1) || (req.mmu_info.els_mmu_flag & 1)) {
-		if (scheduler->data->mmu != RGA_MMU) {
-			rga_job_err(job, "core[%d] has no MMU, please use physically contiguous memory.\n",
-				scheduler->core);
-			rga_job_err(job, "mmu_flag[src, src1, dst, els] = [0x%x, 0x%x, 0x%x, 0x%x]\n",
-				req.mmu_info.src0_mmu_flag, req.mmu_info.src1_mmu_flag,
-				req.mmu_info.dst_mmu_flag, req.mmu_info.els_mmu_flag);
-			return -EINVAL;
+		rga2_align_check(job, &req);
+
+		/* RGA2 mmu set */
+		if ((req.mmu_info.src0_mmu_flag & 1) || (req.mmu_info.src1_mmu_flag & 1) ||
+			(req.mmu_info.dst_mmu_flag & 1) || (req.mmu_info.els_mmu_flag & 1)) {
+			if (scheduler->data->mmu != RGA_MMU) {
+				rga_job_err(job, "core[%d] has no MMU, please use physically contiguous memory.\n",
+					scheduler->core);
+				rga_job_err(job, "mmu_flag[src, src1, dst, els] = [0x%x, 0x%x, 0x%x, 0x%x]\n",
+					req.mmu_info.src0_mmu_flag, req.mmu_info.src1_mmu_flag,
+					req.mmu_info.dst_mmu_flag, req.mmu_info.els_mmu_flag);
+				return -EINVAL;
+			}
+
+			ret = rga_set_mmu_base(job, task_buffers, &req);
+			if (ret < 0) {
+				rga_job_err(job, "%s, [%d] set mmu info error\n", __func__,
+					__LINE__);
+				return -EFAULT;
+			}
 		}
 
-		ret = rga_set_mmu_base(job, &req);
+		/* In slave mode, the current frame completion interrupt must be enabled. */
+		if (scheduler->data->mmu == RGA_IOMMU && job->task_count == 1)
+			req.CMD_fin_int_enable = 1;
+
+		ret = rga2_gen_reg_info(scheduler, (uint8_t *)cmd_base, &req);
 		if (ret < 0) {
-			rga_job_err(job, "%s, [%d] set mmu info error\n", __func__,
-				 __LINE__);
-			return -EFAULT;
+			rga_job_err(job, "gen reg info error\n");
+			return -EINVAL;
 		}
 	}
 
-	/* In slave mode, the current frame completion interrupt must be enabled. */
-	if (scheduler->data->mmu == RGA_IOMMU)
-		req.CMD_fin_int_enable = 1;
+	return 0;
+}
 
-	ret = rga2_gen_reg_info(scheduler, (uint8_t *)job->cmd_buf->vaddr, &req);
-	if (ret < 0) {
-		rga_job_err(job, "gen reg info error\n");
-		return -EINVAL;
-	}
+static int rga2_init_reg(struct rga_job *job)
+{
+	int ret;
+	ktime_t timestamp = ktime_get();
+
+	ret = rga2_init_sys_reg(job);
+	if (ret < 0)
+		return ret;
+
+	ret = rga2_init_cmd_reg(job);
+	if (ret < 0)
+		return ret;
 
 	if (DEBUGGER_EN(TIME))
 		rga_job_log(job, "generate register cost time %lld us\n",
 			ktime_us_delta(ktime_get(), timestamp));
 
-	return ret;
+	return 0;
 }
 
 static void rga2_dump_read_back_sys_reg(struct rga_job *job, struct rga_scheduler_t *scheduler)
@@ -3126,20 +3171,18 @@ static void rga2_set_reg_full_csc(struct rga_job *job, struct rga_scheduler_t *s
 
 static int rga2_set_reg(struct rga_job *job, struct rga_scheduler_t *scheduler)
 {
-	int i;
+	int i, j;
 	bool master_mode_en;
 	uint32_t sys_ctrl;
 	uint32_t *cmd;
 	unsigned long flags;
 	ktime_t now = ktime_get();
 
-	cmd = job->cmd_buf->vaddr;
-
 	/*
 	 * Currently there is no iova allocated for storing cmd for the IOMMU device,
 	 * so the iommu device needs to use the slave mode.
 	 */
-	if (scheduler->data->mmu != RGA_IOMMU)
+	if (scheduler->data->mmu != RGA_IOMMU || job->task_count > 1)
 		master_mode_en = true;
 	else
 		master_mode_en = false;
@@ -3150,12 +3193,16 @@ static int rga2_set_reg(struct rga_job *job, struct rga_scheduler_t *scheduler)
 		if (scheduler->data->version > 0)
 			rga2_dump_read_back_other_reg(job, scheduler);
 
-		rga_job_log(job, "CMD_REG\n");
-		for (i = 0; i < 8; i++)
-			rga_job_log(job, "0x%04x : %.8x %.8x %.8x %.8x\n",
-				RGA2_CMD_REG_BASE + i * 0x10,
-				cmd[0 + i * 4], cmd[1 + i * 4],
-				cmd[2 + i * 4], cmd[3 + i * 4]);
+		for (i = 0; i < job->task_count; i++) {
+			cmd = (uint32_t *)job->cmd_buf->vaddr + i * scheduler->data->cmd_reg_size;
+			rga_job_log(job, "CMD_REG task[%d]\n", i);
+
+			for (j = 0; j < ALIGN_DOWN(scheduler->data->cmd_reg_size, 4) / 4; j++)
+				rga_job_log(job, "0x%04x : %.8x %.8x %.8x %.8x\n",
+					RGA2_CMD_REG_BASE + j * 0x10,
+					cmd[0 + j * 4], cmd[1 + j * 4],
+					cmd[2 + j * 4], cmd[3 + j * 4]);
+		}
 		rga2_dump_read_back_iommu_reg(scheduler->running_job, scheduler);
 	}
 
@@ -3196,14 +3243,19 @@ static int rga2_set_reg(struct rga_job *job, struct rga_scheduler_t *scheduler)
 		/* set cmd_addr */
 		rga_write(job->cmd_buf->dma_addr, RGA2_CMD_BASE, scheduler);
 		rga_write(sys_ctrl, RGA2_SYS_CTRL, scheduler);
-		rga_write(rga_read(RGA2_CMD_CTRL, scheduler) | m_RGA2_CMD_CTRL_CMD_LINE_ST_P,
+
+		rga_write(s_RGA2_MAX_RD_OUTSTANDING(RGA2_MAX_RD_OUTSTANDING) |
+			  s_RGA2_CMD_CTRL_INCR_NUM(job->task_count - 1) |
+			  s_RGA2_CMD_CTRL_STOP(0) |
+			  m_RGA2_CMD_CTRL_CMD_LINE_ST_P,
 			  RGA2_CMD_CTRL, scheduler);
 	} else {
 		/* slave mode */
 		sys_ctrl |= s_RGA2_SYS_CTRL_CMD_MODE(0) | m_RGA2_SYS_CTRL_CMD_OP_ST_P;
 
 		/* set cmd_reg */
-		for (i = 0; i <= 32; i++)
+		cmd = job->cmd_buf->vaddr;
+		for (i = 0; i < scheduler->data->cmd_reg_size; i++)
 			rga_write(cmd[i], 0x100 + i * 4, scheduler);
 
 		rga_write(sys_ctrl, RGA2_SYS_CTRL, scheduler);
@@ -3260,17 +3312,24 @@ static int rga2_get_version(struct rga_scheduler_t *scheduler)
 
 static int rga2_read_back_reg(struct rga_job *job, struct rga_scheduler_t *scheduler)
 {
-	if (job->rga_command_base.osd_info.enable) {
-		if (scheduler->data->version == 0) {
-			job->rga_command_base.osd_info.cur_flags0 =
-				rga_read(RGA2_OSD_CUR_FLAGS0, scheduler);
-			job->rga_command_base.osd_info.cur_flags1 =
-				rga_read(RGA2_OSD_CUR_FLAGS1, scheduler);
-		} else {
-			job->rga_command_base.osd_info.cur_flags0 =
-				rga_read(RGA2_FIXED_OSD_CUR_FLAGS0, scheduler);
-			job->rga_command_base.osd_info.cur_flags1 =
-				rga_read(RGA2_FIXED_OSD_CUR_FLAGS1, scheduler);
+	int i;
+	struct rga_req *user_req = NULL;
+
+	for (i = 0; i < job->task_count; i++) {
+		user_req = &job->task_list[i];
+
+		if (user_req->osd_info.enable) {
+			if (scheduler->data->version == 0) {
+				user_req->osd_info.cur_flags0 =
+					rga_read(RGA2_OSD_CUR_FLAGS0, scheduler);
+				user_req->osd_info.cur_flags1 =
+					rga_read(RGA2_OSD_CUR_FLAGS1, scheduler);
+			} else {
+				user_req->osd_info.cur_flags0 =
+					rga_read(RGA2_FIXED_OSD_CUR_FLAGS0, scheduler);
+				user_req->osd_info.cur_flags1 =
+					rga_read(RGA2_FIXED_OSD_CUR_FLAGS1, scheduler);
+			}
 		}
 	}
 
@@ -3298,6 +3357,8 @@ static void rga2_clear_intr(struct rga_scheduler_t *scheduler)
 
 static int rga2_irq(struct rga_scheduler_t *scheduler)
 {
+	uint32_t cmd_cur_num;
+	uint32_t cmd_ctrl;
 	struct rga_job *job = scheduler->running_job;
 
 	/* The hardware interrupt top-half don't need to lock the scheduler. */
@@ -3318,14 +3379,22 @@ static int rga2_irq(struct rga_scheduler_t *scheduler)
 	}
 
 	scheduler->ops->read_status(job, scheduler);
+	if (v_RGA2_SYS_CTRL_CMD_MODE(rga_read(RGA2_SYS_CTRL, scheduler)) == 0)
+		cmd_cur_num = 1;
+	else
+		cmd_cur_num = v_RGA2_STATUS1_SW_CMD_CUR_NUM(job->cmd_status);
 
-	if (DEBUGGER_EN(INT_FLAG))
+	if (DEBUGGER_EN(INT_FLAG)) {
+		rga_job_log(job, "current cmd %d/%d\n", cmd_cur_num,
+			v_RGA2_STATUS1_SW_CMD_TOTAL_NUM(job->cmd_status) + 1);
 		rga_job_log(job, "irq handler, INTR[0x%x], HW_STATUS[0x%x], CMD_STATUS[0x%x], WORK_CYCLE[0x%x(%d)]\n",
 			job->intr_status, job->hw_status, job->cmd_status,
 			job->work_cycle, job->work_cycle);
+	}
 
 	if (job->intr_status & m_RGA2_INT_ERROR_FLAG_MASK) {
 		set_bit(RGA_JOB_STATE_INTR_ERR, &job->state);
+		job->finished_count = cmd_cur_num > 0 ? cmd_cur_num - 1 : 0;
 
 		rga_job_err(job, "irq handler err! INTR[0x%x], HW_STATUS[0x%x], CMD_STATUS[0x%x], WORK_CYCLE[0x%x(%d)]\n",
 			job->intr_status, job->hw_status, job->cmd_status,
@@ -3334,7 +3403,45 @@ static int rga2_irq(struct rga_scheduler_t *scheduler)
 		scheduler->ops->soft_reset(scheduler);
 	} else if (job->intr_status &
 		   (m_RGA2_INT_CUR_CMD_DONE_INT_FLAG | m_RGA2_INT_ALL_CMD_DONE_INT_FLAG)) {
-		set_bit(RGA_JOB_STATE_FINISH, &job->state);
+		if (job->task_count == 1) {
+			set_bit(RGA_JOB_STATE_FINISH, &job->state);
+			job->finished_count = cmd_cur_num;
+		} else {
+			cmd_ctrl = rga_read(RGA2_CMD_CTRL, scheduler);
+
+			if ((cmd_ctrl & m_RGA2_CMD_CTRL_STOP) > 0 &&
+			    v_RGA2_STATUS1_SW_CMD_CUR_NUM(job->cmd_status) < job->task_count) {
+				/*
+				 *  In addition to continue hardware running,
+				 * INCR_VALID_P will also increment the total
+				 * task count by INCR_NUM.
+				 */
+				cmd_ctrl = ((cmd_ctrl & (~m_RGA2_CMD_CTRL_INCR_NUM)) |
+					   s_RGA2_CMD_CTRL_INCR_NUM(0));
+
+				rga_write(rga_read(RGA2_INT, scheduler) |
+					  m_RGA2_INT_NOW_CMD_DONE_INT_CLEAR |
+					  m_RGA2_INT_ALL_CMD_DONE_INT_CLEAR,
+					  RGA2_INT, scheduler);
+				rga_write(cmd_ctrl |
+					  m_RGA2_CMD_CTRL_INCR_VALID_P,
+					  RGA2_CMD_CTRL, scheduler);
+
+				return IRQ_HANDLED;
+			}
+
+			if ((job->intr_status & m_RGA2_INT_ALL_CMD_DONE_INT_FLAG) == 0) {
+				/* clear current frame done interrupt */
+				rga_write(rga_read(RGA2_INT, scheduler) |
+					  m_RGA2_INT_NOW_CMD_DONE_INT_CLEAR,
+					  RGA2_INT, scheduler);
+
+				return IRQ_HANDLED;
+			}
+
+			set_bit(RGA_JOB_STATE_FINISH, &job->state);
+			job->finished_count = cmd_cur_num;
+		}
 	}
 
 	rga2_clear_intr(scheduler);
diff --git a/drivers/video/rockchip/rga3/rga3_reg_info.c b/drivers/video/rockchip/rga3/rga3_reg_info.c
index 438ce783d09a..fceca4a06791 100644
--- a/drivers/video/rockchip/rga3/rga3_reg_info.c
+++ b/drivers/video/rockchip/rga3/rga3_reg_info.c
@@ -1927,61 +1927,65 @@ static int rga3_check_param(struct rga_job *job, const struct rga_hw_data *data,
 	return 0;
 }
 
-static void print_debug_info(struct rga_job *job, struct rga3_req *req)
+static void print_debug_info(struct rga_job *job, struct rga3_req *req, int task_index)
 {
-	rga_job_log(job, "render_mode:%s, bitblit_mode=%d, rotate_mode:%x\n",
-		rga_get_render_mode_str(req->render_mode), req->bitblt_mode,
+	rga_job_log(job, "task[%d]: render_mode:%s, bitblit_mode=%d, rotate_mode:%x\n",
+		task_index, rga_get_render_mode_str(req->render_mode), req->bitblt_mode,
 		req->rotate_mode);
-	rga_job_log(job, "win0: y = %lx uv = %lx v = %lx src_w = %d src_h = %d\n",
-		req->win0.yrgb_addr, req->win0.uv_addr, req->win0.v_addr,
+	rga_job_log(job, "task[%d]: win0: y = %lx uv = %lx v = %lx src_w = %d src_h = %d\n",
+		task_index, req->win0.yrgb_addr, req->win0.uv_addr, req->win0.v_addr,
 		req->win0.src_act_w, req->win0.src_act_h);
-	rga_job_log(job, "win0: vw = %d vh = %d xoff = %d yoff = %d format = %s\n",
-		req->win0.vir_w, req->win0.vir_h,
+	rga_job_log(job, "task[%d]: win0: vw = %d vh = %d xoff = %d yoff = %d format = %s\n",
+		task_index, req->win0.vir_w, req->win0.vir_h,
 		req->win0.x_offset, req->win0.y_offset,
 		rga_get_format_name(req->win0.format));
-	rga_job_log(job, "win0: dst_w = %d, dst_h = %d, rd_mode = %d\n",
-		req->win0.dst_act_w, req->win0.dst_act_h, req->win0.rd_mode);
-	rga_job_log(job, "win0: rot_mode = %d, en = %d, compact = %d, endian = %d\n",
-		req->win0.rotate_mode, req->win0.enable,
+	rga_job_log(job, "task[%d]: win0: dst_w = %d, dst_h = %d, rd_mode = %d\n",
+		task_index, req->win0.dst_act_w, req->win0.dst_act_h, req->win0.rd_mode);
+	rga_job_log(job, "task[%d]: win0: rot_mode = %d, en = %d, compact = %d, endian = %d\n",
+		task_index, req->win0.rotate_mode, req->win0.enable,
 		req->win0.is_10b_compact, req->win0.is_10b_endian);
 
 	if (req->win1.yrgb_addr != 0 || req->win1.uv_addr != 0
 		|| req->win1.v_addr != 0) {
-		rga_job_log(job, "win1: y = %lx uv = %lx v = %lx src_w = %d src_h = %d\n",
-			req->win1.yrgb_addr, req->win1.uv_addr,
+		rga_job_log(job, "task[%d]: win1: y = %lx uv = %lx v = %lx src_w = %d src_h = %d\n",
+			task_index, req->win1.yrgb_addr, req->win1.uv_addr,
 			req->win1.v_addr, req->win1.src_act_w,
 			req->win1.src_act_h);
-		rga_job_log(job, "win1: vw = %d vh = %d xoff = %d yoff = %d format = %s\n",
-			req->win1.vir_w, req->win1.vir_h,
+		rga_job_log(job, "task[%d]: win1: vw = %d vh = %d xoff = %d yoff = %d format = %s\n",
+			task_index, req->win1.vir_w, req->win1.vir_h,
 			req->win1.x_offset, req->win1.y_offset,
 			rga_get_format_name(req->win1.format));
-		rga_job_log(job, "win1: dst_w = %d, dst_h = %d, rd_mode = %d\n",
-			req->win1.dst_act_w, req->win1.dst_act_h,
+		rga_job_log(job, "task[%d]: win1: dst_w = %d, dst_h = %d, rd_mode = %d\n",
+			task_index, req->win1.dst_act_w, req->win1.dst_act_h,
 			req->win1.rd_mode);
-		rga_job_log(job, "win1: rot_mode = %d, en = %d, compact = %d, endian = %d\n",
-			req->win1.rotate_mode, req->win1.enable,
+		rga_job_log(job, "task[%d]: win1: rot_mode = %d, en = %d, compact = %d, endian = %d\n",
+			task_index, req->win1.rotate_mode, req->win1.enable,
 			req->win1.is_10b_compact, req->win1.is_10b_endian);
 	}
 
-	rga_job_log(job, "wr: y = %lx uv = %lx v = %lx vw = %d vh = %d\n",
-		req->wr.yrgb_addr, req->wr.uv_addr, req->wr.v_addr,
+	rga_job_log(job, "task[%d]: wr: y = %lx uv = %lx v = %lx vw = %d vh = %d\n",
+		task_index, req->wr.yrgb_addr, req->wr.uv_addr, req->wr.v_addr,
 		req->wr.vir_w, req->wr.vir_h);
-	rga_job_log(job, "wr: ovlp_xoff = %d ovlp_yoff = %d format = %s rdmode = %d\n",
-		req->wr.x_offset, req->wr.y_offset,
+	rga_job_log(job, "task[%d]: wr: ovlp_xoff = %d ovlp_yoff = %d format = %s rdmode = %d\n",
+		task_index, req->wr.x_offset, req->wr.y_offset,
 		rga_get_format_name(req->wr.format), req->wr.rd_mode);
 
-	rga_job_log(job, "mmu: win0 = %.2x win1 = %.2x wr = %.2x\n",
+	rga_job_log(job, "task[%d]: mmu: win0 = %.2x win1 = %.2x wr = %.2x\n",
+		task_index,
 		req->mmu_info.src0_mmu_flag, req->mmu_info.src1_mmu_flag,
 		req->mmu_info.dst_mmu_flag);
-	rga_job_log(job, "alpha: flag %x mode=%s\n",
+	rga_job_log(job, "task[%d]: alpha: flag %x mode=%s\n",
+		task_index,
 		req->alpha_rop_flag, rga_get_blend_mode_str(req->alpha_config.mode));
-	rga_job_log(job, "alpha: pre_multi=[%d,%d] pixl=[%d,%d] glb=[%d,%d]\n",
+	rga_job_log(job, "task[%d]: alpha: pre_multi=[%d,%d] pixl=[%d,%d] glb=[%d,%d]\n",
+		task_index,
 		req->alpha_config.fg_pre_multiplied, req->alpha_config.bg_pre_multiplied,
 		req->alpha_config.fg_pixel_alpha_en, req->alpha_config.bg_pixel_alpha_en,
 		req->alpha_config.fg_global_alpha_en, req->alpha_config.bg_global_alpha_en);
-	rga_job_log(job, "alpha: fg_global_alpha=%x bg_global_alpha=%x\n",
+	rga_job_log(job, "task[%d]: alpha: fg_global_alpha=%x bg_global_alpha=%x\n",
+		task_index,
 		req->alpha_config.fg_global_alpha_value, req->alpha_config.bg_global_alpha_value);
-	rga_job_log(job, "yuv2rgb mode is %x\n", req->yuv2rgb_mode);
+	rga_job_log(job, "task[%d]: yuv2rgb mode is %x\n", task_index, req->yuv2rgb_mode);
 }
 
 static int rga3_align_check(struct rga_job *job, struct rga3_req *req)
@@ -2007,8 +2011,11 @@ static int rga3_align_check(struct rga_job *job, struct rga3_req *req)
 
 static int rga3_init_reg(struct rga_job *job)
 {
-	struct rga3_req req;
 	int ret = 0;
+	int i;
+	uint32_t *cmd_base = NULL;
+	struct rga_req *user_req = NULL;
+	struct rga3_req req;
 	struct rga_scheduler_t *scheduler = NULL;
 	ktime_t timestamp = ktime_get();
 
@@ -2018,26 +2025,31 @@ static int rga3_init_reg(struct rga_job *job)
 		return -EINVAL;
 	}
 
-	memset(&req, 0x0, sizeof(req));
+	for (i = 0; i < job->task_count; i++) {
+		user_req = &job->task_list[i];
+		cmd_base = (uint32_t *)job->cmd_buf->vaddr + i * scheduler->data->cmd_reg_size;
 
-	rga_cmd_to_rga3_cmd(&job->rga_command_base, &req);
+		memset(&req, 0x0, sizeof(req));
 
-	/* for debug */
-	if (DEBUGGER_EN(MSG))
-		print_debug_info(job, &req);
+		rga_cmd_to_rga3_cmd(user_req, &req);
 
-	/* check value if legal */
-	ret = rga3_check_param(job, scheduler->data, &req);
-	if (ret == -EINVAL) {
-		rga_job_err(job, "req argument is inval\n");
-		return ret;
-	}
+		/* for debug */
+		if (DEBUGGER_EN(MSG))
+			print_debug_info(job, &req, i);
 
-	rga3_align_check(job, &req);
+		/* check value if legal */
+		ret = rga3_check_param(job, scheduler->data, &req);
+		if (ret == -EINVAL) {
+			rga_job_err(job, "req argument is inval\n");
+			return ret;
+		}
 
-	if (rga3_gen_reg_info((uint8_t *) job->cmd_buf->vaddr, &req) == -1) {
-		rga_job_err(job, "RKA: gen reg info error\n");
-		return -EINVAL;
+		rga3_align_check(job, &req);
+
+		if (rga3_gen_reg_info((uint8_t *) cmd_base, &req) == -1) {
+			rga_job_err(job, "RKA: gen reg info error\n");
+			return -EINVAL;
+		}
 	}
 
 	if (DEBUGGER_EN(TIME))
@@ -2112,19 +2124,18 @@ static void rga3_dump_read_back_iommu_reg(struct rga_job *job, struct rga_schedu
 
 static int rga3_set_reg(struct rga_job *job, struct rga_scheduler_t *scheduler)
 {
-	int i;
+	int i, j;
 	bool master_mode_en;
 	uint32_t sys_ctrl;
+	uint32_t int_en;
 	uint32_t *cmd;
 	ktime_t now = ktime_get();
 
-	cmd = job->cmd_buf->vaddr;
-
 	/*
 	 * Currently there is no iova allocated for storing cmd for the IOMMU device,
 	 * so the iommu device needs to use the slave mode.
 	 */
-	if (scheduler->data->mmu != RGA_IOMMU)
+	if (scheduler->data->mmu != RGA_IOMMU || job->task_count > 1)
 		master_mode_en = true;
 	else
 		master_mode_en = false;
@@ -2132,19 +2143,26 @@ static int rga3_set_reg(struct rga_job *job, struct rga_scheduler_t *scheduler)
 	if (DEBUGGER_EN(REG)) {
 		rga3_dump_read_back_sys_reg(job, scheduler);
 
-		rga_job_log(job, "CMD_REG\n");
-		for (i = 0; i < 12; i++)
-			rga_job_log(job, "0x%04x : %.8x %.8x %.8x %.8x\n",
-				RGA3_CMD_REG_BASE + i * 0x10,
-				cmd[0 + i * 4], cmd[1 + i * 4],
-				cmd[2 + i * 4], cmd[3 + i * 4]);
+		for (i = 0; i < job->task_count; i++) {
+			cmd = (uint32_t *)job->cmd_buf->vaddr + i * scheduler->data->cmd_reg_size;
+			rga_job_log(job, "CMD_REG task[%d]\n", i);
+
+			for (j = 0; j < ALIGN_DOWN(scheduler->data->cmd_reg_size, 4) / 4; j++)
+				rga_job_log(job, "0x%04x : %.8x %.8x %.8x %.8x\n",
+					RGA3_CMD_REG_BASE + j * 0x10,
+					cmd[0 + j * 4], cmd[1 + j * 4],
+					cmd[2 + j * 4], cmd[3 + j * 4]);
+		}
 
 		rga3_dump_read_back_iommu_reg(job, scheduler);
 	}
 
 	/* All CMD finish int */
-	rga_write(m_RGA3_INT_FRM_DONE | m_RGA3_INT_CMD_LINE_FINISH | m_RGA3_INT_ERROR_MASK,
-		  RGA3_INT_EN, scheduler);
+	if (master_mode_en)
+		int_en = m_RGA3_INT_CMD_LINE_FINISH | m_RGA3_INT_ERROR_MASK;
+	else
+		int_en = m_RGA3_INT_FRM_DONE | m_RGA3_INT_CMD_LINE_FINISH | m_RGA3_INT_ERROR_MASK;
+	rga_write(int_en, RGA3_INT_EN, scheduler);
 
 	sys_ctrl = m_RGA3_SYS_CTRL_RGA_LGC_CLK_ON;
 
@@ -2154,12 +2172,16 @@ static int rga3_set_reg(struct rga_job *job, struct rga_scheduler_t *scheduler)
 
 		rga_write(job->cmd_buf->dma_addr, RGA3_CMD_ADDR, scheduler);
 		rga_write(sys_ctrl, RGA3_SYS_CTRL, scheduler);
-		rga_write(m_RGA3_CMD_CTRL_CMD_LINE_ST_P, RGA3_CMD_CTRL, scheduler);
+		rga_write(s_RGA3_CMD_CTRL_CMD_INCR_NUM(job->task_count - 1) |
+			  s_RGA3_CMD_CTRL_CMD_STOP_MODE(0) |
+			  m_RGA3_CMD_CTRL_CMD_LINE_ST_P,
+			  RGA3_CMD_CTRL, scheduler);
 	} else {
 		/* slave mode */
 		sys_ctrl |= s_RGA3_SYS_CTRL_CMD_MODE(0) | m_RGA3_SYS_CTRL_RGA_SART;
 
-		for (i = 0; i <= 50; i++)
+		cmd = job->cmd_buf->vaddr;
+		for (i = 0; i < scheduler->data->cmd_reg_size; i++)
 			rga_write(cmd[i], 0x100 + i * 4, scheduler);
 
 		rga_write(sys_ctrl, RGA3_SYS_CTRL, scheduler);
@@ -2227,6 +2249,7 @@ static void rga3_clear_intr(struct rga_scheduler_t *scheduler)
 
 static int rga3_irq(struct rga_scheduler_t *scheduler)
 {
+	uint32_t cmd_cur_num;
 	struct rga_job *job = scheduler->running_job;
 
 	if (job == NULL) {
@@ -2244,19 +2267,51 @@ static int rga3_irq(struct rga_scheduler_t *scheduler)
 	}
 
 	scheduler->ops->read_status(job, scheduler);
+	if (v_RGA3_SYS_CTRL_CMD_MODE(rga_read(RGA3_SYS_CTRL, scheduler)) == 0)
+		cmd_cur_num = 1;
+	else
+		cmd_cur_num = v_RGA3_CMD_STATE_CMD_CNT_CUR(job->cmd_status);
 
-	if (DEBUGGER_EN(INT_FLAG))
+	if (DEBUGGER_EN(INT_FLAG)) {
+		rga_job_log(job, "current cmd %d/%zu\n", cmd_cur_num, job->task_count);
 		rga_job_log(job, "irq handler, INTR[0x%x], HW_STATUS[0x%x], CMD_STATUS[0x%x]\n",
 			job->intr_status, job->hw_status, job->cmd_status);
+	}
 
-	if (job->intr_status & (m_RGA3_INT_FRM_DONE | m_RGA3_INT_CMD_LINE_FINISH)) {
-		set_bit(RGA_JOB_STATE_FINISH, &job->state);
-	} else if (job->intr_status & m_RGA3_INT_ERROR_MASK) {
+	if (job->intr_status & m_RGA3_INT_ERROR_MASK) {
 		set_bit(RGA_JOB_STATE_INTR_ERR, &job->state);
+		job->finished_count = cmd_cur_num > 0 ? cmd_cur_num - 1 : 0;
 
 		rga_job_err(job, "irq handler err! INTR[0x%x], HW_STATUS[0x%x], CMD_STATUS[0x%x]\n",
 		       job->intr_status, job->hw_status, job->cmd_status);
+
 		scheduler->ops->soft_reset(scheduler);
+	} else if (job->intr_status & (m_RGA3_INT_FRM_DONE | m_RGA3_INT_CMD_LINE_FINISH)) {
+		if (job->task_count == 1) {
+			set_bit(RGA_JOB_STATE_FINISH, &job->state);
+			job->finished_count = cmd_cur_num;
+		} else {
+			if (rga_read(RGA3_CMD_CTRL, scheduler) & m_RGA3_CMD_CTRL_CMD_STOP_MODE &&
+			    v_RGA3_CMD_STATE_CMD_CNT_CUR(job->cmd_status) < job->task_count) {
+				rga_write(m_RGA3_INT_FRM_DONE | m_RGA3_INT_CMD_LINE_FINISH,
+					  RGA3_INT_CLR, scheduler);
+				rga_write(rga_read(RGA3_CMD_CTRL, scheduler) |
+					  m_RGA3_CMD_CTRL_CMD_INCR_VALID_P,
+					  RGA3_CMD_CTRL, scheduler);
+
+				return IRQ_HANDLED;
+			}
+
+			if ((job->intr_status & m_RGA3_INT_CMD_LINE_FINISH) == 0) {
+				/* clear current frame done interrupt */
+				rga_write(m_RGA3_INT_FRM_DONE, RGA3_INT_CLR, scheduler);
+
+				return IRQ_HANDLED;
+			}
+
+			set_bit(RGA_JOB_STATE_FINISH, &job->state);
+			job->finished_count = cmd_cur_num;
+		}
 	}
 
 	rga3_clear_intr(scheduler);
diff --git a/drivers/video/rockchip/rga3/rga_debugger.c b/drivers/video/rockchip/rga3/rga_debugger.c
index 612db0ffb603..59d6464cfff7 100644
--- a/drivers/video/rockchip/rga3/rga_debugger.c
+++ b/drivers/video/rockchip/rga3/rga_debugger.c
@@ -867,6 +867,7 @@ void rga_request_task_debug_info(struct seq_file *m, struct rga_req *req)
 #ifdef CONFIG_NO_GKI
 static int rga_dump_image_to_file(struct rga_internal_buffer *dump_buffer,
 				  const char *channel_name,
+				  int task_index,
 				  int plane_id,
 				  int core)
 {
@@ -938,16 +939,16 @@ static int rga_dump_image_to_file(struct rga_internal_buffer *dump_buffer,
 
 	if (dump_buffer->memory_parm.width == 0 &&
 	    dump_buffer->memory_parm.height == 0)
-		snprintf(file_name, 100, "%s/%d_core%d_%s_plane%d_%s_size%zu_%s.bin",
+		snprintf(file_name, 100, "%s/%d_task%d_core%d_%s_plane%d_%s_size%zu_%s.bin",
 			 g_dump_path,
-			 RGA_DEBUG_DUMP_IMAGE, core, channel_name, plane_id,
+			 RGA_DEBUG_DUMP_IMAGE, task_index, core, channel_name, plane_id,
 			 rga_get_memory_type_str(dump_buffer->type),
 			 size,
 			 rga_get_format_name(dump_buffer->memory_parm.format));
 	else
-		snprintf(file_name, 100, "%s/%d_core%d_%s_plane%d_%s_w%d_h%d_%s.bin",
+		snprintf(file_name, 100, "%s/%d_task%d_core%d_%s_plane%d_%s_w%d_h%d_%s.bin",
 			 g_dump_path,
-			 RGA_DEBUG_DUMP_IMAGE, core, channel_name, plane_id,
+			 RGA_DEBUG_DUMP_IMAGE, task_index, core, channel_name, plane_id,
 			 rga_get_memory_type_str(dump_buffer->type),
 			 dump_buffer->memory_parm.width,
 			 dump_buffer->memory_parm.height,
@@ -981,22 +982,31 @@ static int rga_dump_image_to_file(struct rga_internal_buffer *dump_buffer,
 
 static inline void rga_dump_channel_image(struct rga_job_buffer *job_buffer,
 					  const char *channel_name,
+					  int task_index,
 					  int core)
 {
 	if (job_buffer->y_addr)
-		rga_dump_image_to_file(job_buffer->y_addr, channel_name, 0, core);
+		rga_dump_image_to_file(job_buffer->y_addr, channel_name, task_index, 0, core);
 	if (job_buffer->uv_addr)
-		rga_dump_image_to_file(job_buffer->uv_addr, channel_name, 1, core);
+		rga_dump_image_to_file(job_buffer->uv_addr, channel_name, task_index, 1, core);
 	if (job_buffer->v_addr)
-		rga_dump_image_to_file(job_buffer->v_addr, channel_name, 2, core);
+		rga_dump_image_to_file(job_buffer->v_addr, channel_name, task_index, 2, core);
 }
 
 void rga_dump_job_image(struct rga_job *dump_job)
 {
-	rga_dump_channel_image(&dump_job->src_buffer, "src", dump_job->core);
-	rga_dump_channel_image(&dump_job->src1_buffer, "src1", dump_job->core);
-	rga_dump_channel_image(&dump_job->dst_buffer, "dst", dump_job->core);
-	rga_dump_channel_image(&dump_job->els_buffer, "els", dump_job->core);
+	int i;
+
+	for (i = 0; i < dump_job->task_count; i++) {
+		rga_dump_channel_image(&dump_job->task_buffers[i].src_buffer,
+				       "src", i, dump_job->core);
+		rga_dump_channel_image(&dump_job->task_buffers[i].src1_buffer,
+				       "src1", i, dump_job->core);
+		rga_dump_channel_image(&dump_job->task_buffers[i].dst_buffer,
+				       "dst", i, dump_job->core);
+		rga_dump_channel_image(&dump_job->task_buffers[i].els_buffer,
+				       "els", i, dump_job->core);
+	}
 
 	if (RGA_DEBUG_DUMP_IMAGE > 0)
 		RGA_DEBUG_DUMP_IMAGE--;
diff --git a/drivers/video/rockchip/rga3/rga_iommu.c b/drivers/video/rockchip/rga3/rga_iommu.c
index 00956e28dfd0..69b73bdcb654 100644
--- a/drivers/video/rockchip/rga3/rga_iommu.c
+++ b/drivers/video/rockchip/rga3/rga_iommu.c
@@ -52,30 +52,32 @@ int rga_user_memory_check(struct page **pages, u32 w, u32 h, u32 format, int fla
 	return 0;
 }
 
-int rga_set_mmu_base(struct rga_job *job, struct rga2_req *req)
+int rga_set_mmu_base(struct rga_job *job,
+		     struct rga_job_task_buffers *task_buffers,
+		     struct rga2_req *req)
 {
-	if (job->src_buffer.page_table) {
-		rga_dma_sync_flush_range(job->src_buffer.page_table,
-					 (job->src_buffer.page_table +
-					  job->src_buffer.page_count),
+	if (task_buffers->src_buffer.page_table) {
+		rga_dma_sync_flush_range(task_buffers->src_buffer.page_table,
+					 (task_buffers->src_buffer.page_table +
+					  task_buffers->src_buffer.page_count),
 					 job->scheduler);
-		req->mmu_info.src0_base_addr = virt_to_phys(job->src_buffer.page_table);
+		req->mmu_info.src0_base_addr = virt_to_phys(task_buffers->src_buffer.page_table);
 	}
 
-	if (job->src1_buffer.page_table) {
-		rga_dma_sync_flush_range(job->src1_buffer.page_table,
-					 (job->src1_buffer.page_table +
-					  job->src1_buffer.page_count),
+	if (task_buffers->src1_buffer.page_table) {
+		rga_dma_sync_flush_range(task_buffers->src1_buffer.page_table,
+					 (task_buffers->src1_buffer.page_table +
+					  task_buffers->src1_buffer.page_count),
 					 job->scheduler);
-		req->mmu_info.src1_base_addr = virt_to_phys(job->src1_buffer.page_table);
+		req->mmu_info.src1_base_addr = virt_to_phys(task_buffers->src1_buffer.page_table);
 	}
 
-	if (job->dst_buffer.page_table) {
-		rga_dma_sync_flush_range(job->dst_buffer.page_table,
-					 (job->dst_buffer.page_table +
-					  job->dst_buffer.page_count),
+	if (task_buffers->dst_buffer.page_table) {
+		rga_dma_sync_flush_range(task_buffers->dst_buffer.page_table,
+					 (task_buffers->dst_buffer.page_table +
+					  task_buffers->dst_buffer.page_count),
 					 job->scheduler);
-		req->mmu_info.dst_base_addr = virt_to_phys(job->dst_buffer.page_table);
+		req->mmu_info.dst_base_addr = virt_to_phys(task_buffers->dst_buffer.page_table);
 
 		if (((req->alpha_rop_flag & 1) == 1) && (req->bitblt_mode == 0)) {
 			req->mmu_info.src1_base_addr = req->mmu_info.dst_base_addr;
@@ -83,12 +85,12 @@ int rga_set_mmu_base(struct rga_job *job, struct rga2_req *req)
 		}
 	}
 
-	if (job->els_buffer.page_table) {
-		rga_dma_sync_flush_range(job->els_buffer.page_table,
-					 (job->els_buffer.page_table +
-					  job->els_buffer.page_count),
+	if (task_buffers->els_buffer.page_table) {
+		rga_dma_sync_flush_range(task_buffers->els_buffer.page_table,
+					 (task_buffers->els_buffer.page_table +
+					  task_buffers->els_buffer.page_count),
 					 job->scheduler);
-		req->mmu_info.els_base_addr = virt_to_phys(job->els_buffer.page_table);
+		req->mmu_info.els_base_addr = virt_to_phys(task_buffers->els_buffer.page_table);
 	}
 
 	return 0;
diff --git a/drivers/video/rockchip/rga3/rga_job.c b/drivers/video/rockchip/rga3/rga_job.c
index 47d2034af90d..019c308591c1 100644
--- a/drivers/video/rockchip/rga3/rga_job.c
+++ b/drivers/video/rockchip/rga3/rga_job.c
@@ -15,8 +15,16 @@
 
 static void rga_job_free(struct rga_job *job)
 {
-	if (job->cmd_buf)
-		rga_dma_buf_pool_free(job->scheduler->cmd_buf_pool, job->cmd_buf);
+	if (job->cmd_buf) {
+		if (job->task_count > 1)
+			rga_dma_free(job->cmd_buf);
+		else
+			rga_dma_buf_pool_free(job->scheduler->cmd_buf_pool, job->cmd_buf);
+		job->cmd_buf = NULL;
+	}
+
+	kfree(job->task_buffers);
+	job->task_buffers = NULL;
 
 	kfree(job);
 }
@@ -47,14 +55,12 @@ static int rga_job_cleanup(struct rga_job *job)
 	return 0;
 }
 
-static int rga_job_judgment_support_core(struct rga_job *job)
+static int rga_job_judgment_support_core(struct rga_job *job, struct rga_req *req)
 {
 	int ret = 0;
 	uint32_t mm_flag;
-	struct rga_req *req;
 	struct rga_mm *mm;
 
-	req = &job->rga_command_base;
 	mm = rga_drvdata->mm;
 	if (mm == NULL) {
 		rga_job_err(job, "rga mm is null!\n");
@@ -108,8 +114,9 @@ static int rga_job_judgment_support_core(struct rga_job *job)
 	return ret;
 }
 
-static struct rga_job *rga_job_alloc(struct rga_req *rga_command_base)
+static struct rga_job *rga_job_alloc(struct rga_req *task_list, size_t task_count)
 {
+	int i;
 	struct rga_job *job = NULL;
 
 	job = kzalloc(sizeof(*job), GFP_KERNEL);
@@ -122,13 +129,16 @@ static struct rga_job *rga_job_alloc(struct rga_req *rga_command_base)
 	job->timestamp.init = ktime_get();
 	job->pid = current->pid;
 
-	job->rga_command_base = *rga_command_base;
+	job->task_list = task_list;
+	job->task_count = task_count;
 
-	if (rga_command_base->priority > 0) {
-		if (rga_command_base->priority > RGA_SCHED_PRIORITY_MAX)
-			job->priority = RGA_SCHED_PRIORITY_MAX;
-		else
-			job->priority = rga_command_base->priority;
+	for (i = 0; i < task_count; i++) {
+		if (task_list[i].priority > 0) {
+			if (task_list[i].priority > RGA_SCHED_PRIORITY_MAX)
+				job->priority = RGA_SCHED_PRIORITY_MAX;
+			else if (task_list[i].priority > job->priority)
+				job->priority = task_list[i].priority;
+		}
 	}
 
 	if (DEBUGGER_EN(INTERNAL_MODE)) {
@@ -138,10 +148,12 @@ static struct rga_job *rga_job_alloc(struct rga_req *rga_command_base)
 		return job;
 	}
 
-	if (job->rga_command_base.handle_flag & 1) {
-		job->flags |= RGA_JOB_USE_HANDLE;
-
-		rga_job_judgment_support_core(job);
+	for (i = 0; i < task_count; i++) {
+		if (task_list[i].handle_flag & 1) {
+			job->flags |= RGA_JOB_USE_HANDLE;
+			rga_job_judgment_support_core(job, &task_list[i]);
+			break;
+		}
 	}
 
 	return job;
@@ -401,19 +413,18 @@ static struct rga_scheduler_t *rga_job_schedule(struct rga_job *job)
 	return scheduler;
 }
 
-int rga_job_commit(struct rga_req *rga_command_base, struct rga_request *request)
+int rga_job_commit(struct rga_req *task_list, size_t task_count, struct rga_request *request)
 {
 	int ret;
 	struct rga_job *job = NULL;
 	struct rga_scheduler_t *scheduler = NULL;
 
-	job = rga_job_alloc(rga_command_base);
+	job = rga_job_alloc(task_list, task_count);
 	if (!job) {
 		rga_err("failed to alloc rga job!\n");
 		return -ENOMEM;
 	}
 
-	job->use_batch_mode = request->use_batch_mode;
 	job->request_id = request->id;
 	job->session = request->session;
 	job->mm = request->current_mm;
@@ -424,18 +435,36 @@ int rga_job_commit(struct rga_req *rga_command_base, struct rga_request *request
 		goto err_free_job;
 	}
 
-	job->cmd_buf = rga_dma_buf_pool_alloc(scheduler->cmd_buf_pool);
-	if (job->cmd_buf == NULL) {
-		rga_job_err(job, "failed to alloc command buffer.\n");
-		ret = -ENOMEM;
-		goto err_free_job;
-	}
-
 	/* Memory mapping needs to keep pd enabled. */
 	ret = rga_power_enable(scheduler);
 	if (ret < 0) {
 		rga_job_err(job, "power enable failed");
-		goto err_free_cmd_buf;
+		goto err_free_job;
+	}
+
+	if (job->task_count > 1) {
+		job->cmd_buf = rga_dma_alloc_coherent(job->scheduler,
+			job->task_count * scheduler->data->cmd_reg_size * sizeof(uint32_t));
+		if (job->cmd_buf == NULL) {
+			rga_job_err(job, "Failed to allocate coherent memory for multi-task.\n");
+			ret = -ENOMEM;
+			goto err_power_disable;
+		}
+	} else {
+		job->cmd_buf = rga_dma_buf_pool_alloc(scheduler->cmd_buf_pool);
+		if (job->cmd_buf == NULL) {
+			rga_job_err(job, "failed to alloc command buffer.\n");
+			ret = -ENOMEM;
+			goto err_power_disable;
+		}
+	}
+
+	job->task_buffers =
+		kzalloc(sizeof(struct rga_job_task_buffers) * job->task_count, GFP_KERNEL);
+	if (!job->task_buffers) {
+		rga_job_err(job, "Failed to allocate memory for channel buffers.\n");
+		ret = -ENOMEM;
+		goto err_power_disable;
 	}
 
 	ret = rga_mm_map_job_info(job);
@@ -464,10 +493,6 @@ int rga_job_commit(struct rga_req *rga_command_base, struct rga_request *request
 err_power_disable:
 	rga_power_disable(scheduler);
 
-err_free_cmd_buf:
-	rga_dma_buf_pool_free(scheduler->cmd_buf_pool, job->cmd_buf);
-	job->cmd_buf = NULL;
-
 err_free_job:
 	rga_job_free(job);
 
@@ -752,7 +777,7 @@ static int rga_request_scheduler_job_abort(struct rga_request *request)
 				list_move(&job->head, &list_to_free);
 				scheduler->job_count--;
 
-				todo_abort_count++;
+				todo_abort_count += job->task_count;
 			}
 		}
 
@@ -775,7 +800,7 @@ static int rga_request_scheduler_job_abort(struct rga_request *request)
 
 				rga_req_err(request, "reset core[%d] by request abort",
 					scheduler->core);
-				running_abort_count++;
+				running_abort_count += job->task_count - job->finished_count;
 			}
 		}
 
@@ -923,23 +948,42 @@ int rga_request_commit(struct rga_request *request)
 	int ret;
 	int i = 0;
 
-	if (DEBUGGER_EN(MSG))
+	if (DEBUGGER_EN(MSG)) {
 		rga_req_log(request, "commit process: %s\n", request->session->pname);
+		rga_req_log(request, "flags = %#x, task_count = %d\n",
+			request->flags, request->task_count);
+	}
 
-	for (i = 0; i < request->task_count; i++) {
-		struct rga_req *req = &(request->task_list[i]);
-
+	if (request->flags & RGA_REQUEST_FLAGS_EXEC_SEQUENTIAL) {
 		if (DEBUGGER_EN(MSG)) {
-			rga_req_log(request, "commit task[%d]:\n", i);
-			rga_dump_req(request, req);
+			for (i = 0; i < request->task_count; i++) {
+				rga_req_log(request, "commit task[%d]:\n", i);
+				rga_dump_req(request, &(request->task_list[i]));
+			}
 		}
 
-		ret = rga_job_commit(req, request);
+		ret = rga_job_commit(request->task_list, request->task_count, request);
 		if (ret < 0) {
-			rga_req_err(request, "task[%d] job_commit failed.\n", i);
+			rga_req_err(request, "task_list job_commit failed.\n");
 
 			return ret;
 		}
+	} else {
+		for (i = 0; i < request->task_count; i++) {
+			struct rga_req *req = &(request->task_list[i]);
+
+			if (DEBUGGER_EN(MSG)) {
+				rga_req_log(request, "commit task[%d]:\n", i);
+				rga_dump_req(request, req);
+			}
+
+			ret = rga_job_commit(req, 1, request);
+			if (ret < 0) {
+				rga_req_err(request, "task[%d] job_commit failed.\n", i);
+
+				return ret;
+			}
+		}
 	}
 
 	if (request->sync_mode == RGA_BLIT_SYNC) {
@@ -1021,12 +1065,10 @@ int rga_request_release_signal(struct rga_scheduler_t *scheduler, struct rga_job
 
 	spin_lock_irqsave(&request->lock, flags);
 
-	if (job->ret < 0) {
-		request->failed_task_count++;
+	request->failed_task_count += job->task_count - job->finished_count;
+	request->finished_task_count += job->finished_count;
+	if (job->ret < 0)
 		request->ret = job->ret;
-	} else {
-		request->finished_task_count++;
-	}
 
 	failed_count = request->failed_task_count;
 	finished_count = request->finished_task_count;
@@ -1127,7 +1169,6 @@ struct rga_request *rga_request_config(struct rga_user_request *user_request)
 
 	spin_lock_irqsave(&request->lock, flags);
 
-	request->use_batch_mode = true;
 	request->task_list = task_list;
 	request->task_count = user_request->task_num;
 	request->sync_mode = user_request->sync_mode;
@@ -1187,7 +1228,6 @@ struct rga_request *rga_request_kernel_config(struct rga_user_request *user_requ
 
 	spin_lock_irqsave(&request->lock, flags);
 
-	request->use_batch_mode = true;
 	request->task_list = task_list;
 	request->task_count = user_request->task_num;
 	request->sync_mode = user_request->sync_mode;
@@ -1353,7 +1393,7 @@ int rga_request_mpi_submit(struct rga_req *req, struct rga_request *request)
 	rga_request_get(request);
 	mutex_unlock(&request_manager->lock);
 
-	ret = rga_job_commit(req, request);
+	ret = rga_job_commit(req, 1, request);
 	if (ret < 0) {
 		rga_req_err(request, "failed to commit job!\n");
 		goto err_abort_request;
diff --git a/drivers/video/rockchip/rga3/rga_mm.c b/drivers/video/rockchip/rga3/rga_mm.c
index 9ab9e7f60fa8..cdc9ae47d7f6 100644
--- a/drivers/video/rockchip/rga3/rga_mm.c
+++ b/drivers/video/rockchip/rga3/rga_mm.c
@@ -1137,7 +1137,9 @@ static bool rga_mm_is_need_mmu(struct rga_job *job, struct rga_internal_buffer *
 	return false;
 }
 
-static int rga_mm_set_mmu_flag(struct rga_job *job)
+static int rga_mm_set_mmu_flag(struct rga_job *job,
+			       struct rga_job_task_buffers *buffers,
+			       struct rga_req *req)
 {
 	struct rga_mmu_t *mmu_info;
 	int src_mmu_en;
@@ -1145,12 +1147,12 @@ static int rga_mm_set_mmu_flag(struct rga_job *job)
 	int dst_mmu_en;
 	int els_mmu_en;
 
-	src_mmu_en = rga_mm_is_need_mmu(job, job->src_buffer.addr);
-	src1_mmu_en = rga_mm_is_need_mmu(job, job->src1_buffer.addr);
-	dst_mmu_en = rga_mm_is_need_mmu(job, job->dst_buffer.addr);
-	els_mmu_en = rga_mm_is_need_mmu(job, job->els_buffer.addr);
+	src_mmu_en = rga_mm_is_need_mmu(job, buffers->src_buffer.addr);
+	src1_mmu_en = rga_mm_is_need_mmu(job, buffers->src1_buffer.addr);
+	dst_mmu_en = rga_mm_is_need_mmu(job, buffers->dst_buffer.addr);
+	els_mmu_en = rga_mm_is_need_mmu(job, buffers->els_buffer.addr);
 
-	mmu_info = &job->rga_command_base.mmu_info;
+	mmu_info = &req->mmu_info;
 	memset(mmu_info, 0x0, sizeof(*mmu_info));
 	if (src_mmu_en)
 		mmu_info->mmu_flag |= (0x1 << 8);
@@ -1718,14 +1720,14 @@ static int rga_mm_get_channel_handle_info(struct rga_mm *mm,
 	return 0;
 }
 
-static int rga_mm_get_handle_info(struct rga_job *job)
+static int rga_mm_get_handle_info(struct rga_job *job,
+				  struct rga_job_task_buffers *buffers,
+				  struct rga_req *req)
 {
 	int ret = 0;
-	struct rga_req *req = NULL;
 	struct rga_mm *mm = NULL;
 	enum dma_data_direction dir;
 
-	req = &job->rga_command_base;
 	mm = rga_drvdata->mm;
 
 	switch (req->render_mode) {
@@ -1777,7 +1779,7 @@ static int rga_mm_get_handle_info(struct rga_job *job)
 
 	if (likely(req->src.yrgb_addr > 0)) {
 		ret = rga_mm_get_channel_handle_info(mm, job, &req->src,
-						     &job->src_buffer,
+						     &buffers->src_buffer,
 						     DMA_TO_DEVICE);
 		if (ret < 0) {
 			rga_job_err(job, "Can't get src buffer info from handle!\n");
@@ -1787,7 +1789,7 @@ static int rga_mm_get_handle_info(struct rga_job *job)
 
 	if (likely(req->dst.yrgb_addr > 0)) {
 		ret = rga_mm_get_channel_handle_info(mm, job, &req->dst,
-						     &job->dst_buffer,
+						     &buffers->dst_buffer,
 						     DMA_TO_DEVICE);
 		if (ret < 0) {
 			rga_job_err(job, "Can't get dst buffer info from handle!\n");
@@ -1796,7 +1798,6 @@ static int rga_mm_get_handle_info(struct rga_job *job)
 	}
 
 	if (likely(req->pat.yrgb_addr > 0)) {
-
 		if (req->render_mode != UPDATE_PALETTE_TABLE_MODE) {
 			if (req->bsfilter_flag)
 				dir = DMA_BIDIRECTIONAL;
@@ -1804,11 +1805,11 @@ static int rga_mm_get_handle_info(struct rga_job *job)
 				dir = DMA_TO_DEVICE;
 
 			ret = rga_mm_get_channel_handle_info(mm, job, &req->pat,
-							     &job->src1_buffer,
+							     &buffers->src1_buffer,
 							     dir);
 		} else {
 			ret = rga_mm_get_channel_handle_info(mm, job, &req->pat,
-							     &job->els_buffer,
+							     &buffers->els_buffer,
 							     DMA_BIDIRECTIONAL);
 		}
 		if (ret < 0) {
@@ -1817,19 +1818,21 @@ static int rga_mm_get_handle_info(struct rga_job *job)
 		}
 	}
 
-	rga_mm_set_mmu_flag(job);
+	rga_mm_set_mmu_flag(job, buffers, req);
 
 	return 0;
 }
 
-static void rga_mm_put_handle_info(struct rga_job *job)
+static void rga_mm_put_handle_info(struct rga_job *job,
+				   struct rga_job_task_buffers *buffers,
+				   struct rga_req *req)
 {
 	struct rga_mm *mm = rga_drvdata->mm;
 
-	rga_mm_put_channel_handle_info(mm, job, &job->src_buffer, DMA_NONE);
-	rga_mm_put_channel_handle_info(mm, job, &job->dst_buffer, DMA_FROM_DEVICE);
-	rga_mm_put_channel_handle_info(mm, job, &job->src1_buffer, DMA_NONE);
-	rga_mm_put_channel_handle_info(mm, job, &job->els_buffer, DMA_NONE);
+	rga_mm_put_channel_handle_info(mm, job, &buffers->src_buffer, DMA_NONE);
+	rga_mm_put_channel_handle_info(mm, job, &buffers->dst_buffer, DMA_FROM_DEVICE);
+	rga_mm_put_channel_handle_info(mm, job, &buffers->src1_buffer, DMA_NONE);
+	rga_mm_put_channel_handle_info(mm, job, &buffers->els_buffer, DMA_NONE);
 }
 
 static void rga_mm_put_channel_external_buffer(struct rga_job_buffer *job_buffer)
@@ -1886,19 +1889,21 @@ static int rga_mm_get_channel_external_buffer(int mmu_flag,
 	return 0;
 }
 
-static void rga_mm_put_external_buffer(struct rga_job *job)
+static void rga_mm_put_external_buffer(struct rga_job *job, struct rga_job_task_buffers *buffers)
 {
-	if (job->src_buffer.ex_addr)
-		rga_mm_put_channel_external_buffer(&job->src_buffer);
-	if (job->src1_buffer.ex_addr)
-		rga_mm_put_channel_external_buffer(&job->src1_buffer);
-	if (job->dst_buffer.ex_addr)
-		rga_mm_put_channel_external_buffer(&job->dst_buffer);
-	if (job->els_buffer.ex_addr)
-		rga_mm_put_channel_external_buffer(&job->els_buffer);
+	if (buffers->src_buffer.ex_addr)
+		rga_mm_put_channel_external_buffer(&buffers->src_buffer);
+	if (buffers->src1_buffer.ex_addr)
+		rga_mm_put_channel_external_buffer(&buffers->src1_buffer);
+	if (buffers->dst_buffer.ex_addr)
+		rga_mm_put_channel_external_buffer(&buffers->dst_buffer);
+	if (buffers->els_buffer.ex_addr)
+		rga_mm_put_channel_external_buffer(&buffers->els_buffer);
 }
 
-static int rga_mm_get_external_buffer(struct rga_job *job)
+static int rga_mm_get_external_buffer(struct rga_job *job,
+				      struct rga_job_task_buffers *buffers,
+				      struct rga_req *req)
 {
 	int ret = -EINVAL;
 	int mmu_flag;
@@ -1908,20 +1913,20 @@ static int rga_mm_get_external_buffer(struct rga_job *job)
 	struct rga_img_info_t *dst = NULL;
 	struct rga_img_info_t *els = NULL;
 
-	if (job->rga_command_base.render_mode != COLOR_FILL_MODE)
-		src0 = &job->rga_command_base.src;
+	if (req->render_mode != COLOR_FILL_MODE)
+		src0 = &req->src;
 
-	if (job->rga_command_base.render_mode != UPDATE_PALETTE_TABLE_MODE)
-		src1 = job->rga_command_base.bsfilter_flag ?
-		       &job->rga_command_base.pat : NULL;
+	if (req->render_mode != UPDATE_PALETTE_TABLE_MODE)
+		src1 = req->bsfilter_flag ?
+		       &req->pat : NULL;
 	else
-		els = &job->rga_command_base.pat;
+		els = &req->pat;
 
-	dst = &job->rga_command_base.dst;
+	dst = &req->dst;
 
 	if (likely(src0)) {
-		mmu_flag = ((job->rga_command_base.mmu_info.mmu_flag >> 8) & 1);
-		ret = rga_mm_get_channel_external_buffer(mmu_flag, src0, &job->src_buffer);
+		mmu_flag = ((req->mmu_info.mmu_flag >> 8) & 1);
+		ret = rga_mm_get_channel_external_buffer(mmu_flag, src0, &buffers->src_buffer);
 		if (ret < 0) {
 			rga_job_err(job, "Cannot get src0 channel buffer!\n");
 			return ret;
@@ -1929,8 +1934,8 @@ static int rga_mm_get_external_buffer(struct rga_job *job)
 	}
 
 	if (likely(dst)) {
-		mmu_flag = ((job->rga_command_base.mmu_info.mmu_flag >> 10) & 1);
-		ret = rga_mm_get_channel_external_buffer(mmu_flag, dst, &job->dst_buffer);
+		mmu_flag = ((req->mmu_info.mmu_flag >> 10) & 1);
+		ret = rga_mm_get_channel_external_buffer(mmu_flag, dst, &buffers->dst_buffer);
 		if (ret < 0) {
 			rga_job_err(job, "Cannot get dst channel buffer!\n");
 			goto error_put_buffer;
@@ -1938,8 +1943,8 @@ static int rga_mm_get_external_buffer(struct rga_job *job)
 	}
 
 	if (src1) {
-		mmu_flag = ((job->rga_command_base.mmu_info.mmu_flag >> 9) & 1);
-		ret = rga_mm_get_channel_external_buffer(mmu_flag, src1, &job->src1_buffer);
+		mmu_flag = ((req->mmu_info.mmu_flag >> 9) & 1);
+		ret = rga_mm_get_channel_external_buffer(mmu_flag, src1, &buffers->src1_buffer);
 		if (ret < 0) {
 			rga_job_err(job, "Cannot get src1 channel buffer!\n");
 			goto error_put_buffer;
@@ -1947,8 +1952,8 @@ static int rga_mm_get_external_buffer(struct rga_job *job)
 	}
 
 	if (els) {
-		mmu_flag = ((job->rga_command_base.mmu_info.mmu_flag >> 11) & 1);
-		ret = rga_mm_get_channel_external_buffer(mmu_flag, els, &job->els_buffer);
+		mmu_flag = ((req->mmu_info.mmu_flag >> 11) & 1);
+		ret = rga_mm_get_channel_external_buffer(mmu_flag, els, &buffers->els_buffer);
 		if (ret < 0) {
 			rga_job_err(job, "Cannot get els channel buffer!\n");
 			goto error_put_buffer;
@@ -1957,7 +1962,7 @@ static int rga_mm_get_external_buffer(struct rga_job *job)
 
 	return 0;
 error_put_buffer:
-	rga_mm_put_external_buffer(job);
+	rga_mm_put_external_buffer(job, buffers);
 	return ret;
 }
 
@@ -2046,37 +2051,35 @@ static int rga_mm_map_channel_job_buffer(struct rga_job *job,
 	return ret;
 }
 
-static void rga_mm_unmap_buffer_info(struct rga_job *job)
+static void rga_mm_unmap_buffer_info(struct rga_job *job, struct rga_job_task_buffers *buffers)
 {
-	if (job->src_buffer.addr)
-		rga_mm_unmap_channel_job_buffer(job, &job->src_buffer, DMA_NONE);
-	if (job->dst_buffer.addr)
-		rga_mm_unmap_channel_job_buffer(job, &job->dst_buffer, DMA_FROM_DEVICE);
-	if (job->src1_buffer.addr)
-		rga_mm_unmap_channel_job_buffer(job, &job->src1_buffer, DMA_NONE);
-	if (job->els_buffer.addr)
-		rga_mm_unmap_channel_job_buffer(job, &job->els_buffer, DMA_NONE);
-
-	rga_mm_put_external_buffer(job);
+	if (buffers->src_buffer.addr)
+		rga_mm_unmap_channel_job_buffer(job, &buffers->src_buffer, DMA_NONE);
+	if (buffers->dst_buffer.addr)
+		rga_mm_unmap_channel_job_buffer(job, &buffers->dst_buffer, DMA_FROM_DEVICE);
+	if (buffers->src1_buffer.addr)
+		rga_mm_unmap_channel_job_buffer(job, &buffers->src1_buffer, DMA_NONE);
+	if (buffers->els_buffer.addr)
+		rga_mm_unmap_channel_job_buffer(job, &buffers->els_buffer, DMA_NONE);
+	rga_mm_put_external_buffer(job, buffers);
 }
 
-static int rga_mm_map_buffer_info(struct rga_job *job)
+static int rga_mm_map_buffer_info(struct rga_job *job,
+				  struct rga_job_task_buffers *buffers,
+				  struct rga_req *req)
 {
 	int ret = 0;
-	struct rga_req *req = NULL;
 	enum dma_data_direction dir;
 
-	ret = rga_mm_get_external_buffer(job);
+	ret = rga_mm_get_external_buffer(job, buffers, req);
 	if (ret < 0) {
 		rga_job_err(job, "failed to get external buffer from job_cmd!\n");
 		return ret;
 	}
 
-	req = &job->rga_command_base;
-
-	if (likely(job->src_buffer.ex_addr)) {
+	if (likely(buffers->src_buffer.ex_addr)) {
 		ret = rga_mm_map_channel_job_buffer(job, &req->src,
-						    &job->src_buffer,
+						    &buffers->src_buffer,
 						    DMA_TO_DEVICE, false);
 		if (ret < 0) {
 			rga_job_err(job, "src channel map job buffer failed!");
@@ -2084,9 +2087,9 @@ static int rga_mm_map_buffer_info(struct rga_job *job)
 		}
 	}
 
-	if (likely(job->dst_buffer.ex_addr)) {
+	if (likely(buffers->dst_buffer.ex_addr)) {
 		ret = rga_mm_map_channel_job_buffer(job, &req->dst,
-						    &job->dst_buffer,
+						    &buffers->dst_buffer,
 						    DMA_TO_DEVICE, true);
 		if (ret < 0) {
 			rga_job_err(job, "dst channel map job buffer failed!");
@@ -2094,14 +2097,14 @@ static int rga_mm_map_buffer_info(struct rga_job *job)
 		}
 	}
 
-	if (job->src1_buffer.ex_addr) {
+	if (buffers->src1_buffer.ex_addr) {
 		if (req->bsfilter_flag)
 			dir = DMA_BIDIRECTIONAL;
 		else
 			dir = DMA_TO_DEVICE;
 
 		ret = rga_mm_map_channel_job_buffer(job, &req->pat,
-						    &job->src1_buffer,
+						    &buffers->src1_buffer,
 						    dir, false);
 		if (ret < 0) {
 			rga_job_err(job, "src1 channel map job buffer failed!");
@@ -2109,9 +2112,9 @@ static int rga_mm_map_buffer_info(struct rga_job *job)
 		}
 	}
 
-	if (job->els_buffer.ex_addr) {
+	if (buffers->els_buffer.ex_addr) {
 		ret = rga_mm_map_channel_job_buffer(job, &req->pat,
-						    &job->els_buffer,
+						    &buffers->els_buffer,
 						    DMA_BIDIRECTIONAL, false);
 		if (ret < 0) {
 			rga_job_err(job, "els channel map job buffer failed!");
@@ -2119,11 +2122,11 @@ static int rga_mm_map_buffer_info(struct rga_job *job)
 		}
 	}
 
-	rga_mm_set_mmu_flag(job);
+	rga_mm_set_mmu_flag(job, buffers, req);
 	return 0;
 
 error_unmap_buffer:
-	rga_mm_unmap_buffer_info(job);
+	rga_mm_unmap_buffer_info(job, buffers);
 
 	return ret;
 }
@@ -2233,29 +2236,28 @@ static int rga_mm_alloc_channel_fake_buffer(struct rga_job *job,
 	return ret;
 }
 
-static void rga_mm_free_fake_buffer(struct rga_job *job)
+static void rga_mm_free_fake_buffer(struct rga_job *job, struct rga_job_task_buffers *buffers)
 {
-	if (job->src_buffer.addr)
-		rga_mm_free_channel_fake_buffer(job, &job->src_buffer, DMA_NONE);
-	if (job->dst_buffer.addr)
-		rga_mm_free_channel_fake_buffer(job, &job->dst_buffer, DMA_FROM_DEVICE);
-	if (job->src1_buffer.addr)
-		rga_mm_free_channel_fake_buffer(job, &job->src1_buffer, DMA_NONE);
-	if (job->els_buffer.addr)
-		rga_mm_free_channel_fake_buffer(job, &job->els_buffer, DMA_NONE);
+	if (buffers->src_buffer.addr)
+		rga_mm_free_channel_fake_buffer(job, &buffers->src_buffer, DMA_NONE);
+	if (buffers->dst_buffer.addr)
+		rga_mm_free_channel_fake_buffer(job, &buffers->dst_buffer, DMA_FROM_DEVICE);
+	if (buffers->src1_buffer.addr)
+		rga_mm_free_channel_fake_buffer(job, &buffers->src1_buffer, DMA_NONE);
+	if (buffers->els_buffer.addr)
+		rga_mm_free_channel_fake_buffer(job, &buffers->els_buffer, DMA_NONE);
 }
 
-static int rga_mm_alloc_fake_buffer(struct rga_job *job)
+static int rga_mm_alloc_fake_buffer(struct rga_job *job,
+				    struct rga_job_task_buffers *buffers,
+				    struct rga_req *req)
 {
 	int ret = 0;
-	struct rga_req *req = NULL;
 	enum dma_data_direction dir;
 
-	req = &job->rga_command_base;
-
 	if (req->src.yrgb_addr != 0 || req->src.uv_addr != 0) {
 		ret = rga_mm_alloc_channel_fake_buffer(job, &req->src,
-						       &job->src_buffer, DMA_TO_DEVICE);
+						       &buffers->src_buffer, DMA_TO_DEVICE);
 		if (ret < 0) {
 			rga_job_err(job, "%s src channel map job buffer failed!", __func__);
 			goto error_free_fake_buffer;
@@ -2264,14 +2266,14 @@ static int rga_mm_alloc_fake_buffer(struct rga_job *job)
 
 	if (req->dst.yrgb_addr != 0 || req->dst.uv_addr != 0) {
 		ret = rga_mm_alloc_channel_fake_buffer(job, &req->dst,
-						       &job->dst_buffer, DMA_TO_DEVICE);
+						       &buffers->dst_buffer, DMA_TO_DEVICE);
 		if (ret < 0) {
 			rga_job_err(job, "%s dst channel map job buffer failed!", __func__);
 			goto error_free_fake_buffer;
 		}
 	}
 
-	if (job->rga_command_base.render_mode != UPDATE_PALETTE_TABLE_MODE &&
+	if (req->render_mode != UPDATE_PALETTE_TABLE_MODE &&
 	    (req->pat.yrgb_addr != 0 || req->pat.uv_addr != 0)) {
 		if (req->bsfilter_flag)
 			dir = DMA_BIDIRECTIONAL;
@@ -2279,37 +2281,39 @@ static int rga_mm_alloc_fake_buffer(struct rga_job *job)
 			dir = DMA_TO_DEVICE;
 
 		ret = rga_mm_alloc_channel_fake_buffer(job, &req->pat,
-						       &job->src1_buffer, dir);
+						       &buffers->src1_buffer, dir);
 		if (ret < 0) {
 			rga_job_err(job, "%s src1 channel map job buffer failed!", __func__);
 			goto error_free_fake_buffer;
 		}
 	} else if (req->pat.yrgb_addr != 0 || req->pat.uv_addr != 0) {
 		ret = rga_mm_alloc_channel_fake_buffer(job, &req->pat,
-						       &job->els_buffer, DMA_TO_DEVICE);
+						       &buffers->els_buffer, DMA_TO_DEVICE);
 		if (ret < 0) {
 			rga_job_err(job, "%s els channel map job buffer failed!", __func__);
 			goto error_free_fake_buffer;
 		}
 	}
 
-	rga_mm_set_mmu_flag(job);
+	rga_mm_set_mmu_flag(job, buffers, req);
 
 	return 0;
 
 error_free_fake_buffer:
-	rga_mm_free_fake_buffer(job);
+	rga_mm_free_fake_buffer(job, buffers);
 
 	return ret;
 }
 
-int rga_mm_map_job_info(struct rga_job *job)
+static int rga_mm_map_task_info(struct rga_job *job,
+				struct rga_job_task_buffers *buffers,
+				struct rga_req *req)
 {
 	int ret;
 	ktime_t timestamp = ktime_get();
 
 	if (job->flags & RGA_JOB_DEBUG_FAKE_BUFFER) {
-		ret = rga_mm_alloc_fake_buffer(job);
+		ret = rga_mm_alloc_fake_buffer(job, buffers, req);
 		if (ret < 0)
 			return ret;
 
@@ -2324,7 +2328,7 @@ int rga_mm_map_job_info(struct rga_job *job)
 	}
 
 	if (job->flags & RGA_JOB_USE_HANDLE) {
-		ret = rga_mm_get_handle_info(job);
+		ret = rga_mm_get_handle_info(job, buffers, req);
 		if (ret < 0) {
 			rga_job_err(job, "failed to get buffer from handle\n");
 			return ret;
@@ -2334,7 +2338,7 @@ int rga_mm_map_job_info(struct rga_job *job)
 			rga_job_log(job, "get buffer_handle info cost %lld us\n",
 				ktime_us_delta(ktime_get(), timestamp));
 	} else {
-		ret = rga_mm_map_buffer_info(job);
+		ret = rga_mm_map_buffer_info(job, buffers, req);
 		if (ret < 0) {
 			rga_job_err(job, "failed to map buffer\n");
 			return ret;
@@ -2348,12 +2352,14 @@ int rga_mm_map_job_info(struct rga_job *job)
 	return 0;
 }
 
-void rga_mm_unmap_job_info(struct rga_job *job)
+static void rga_mm_unmap_task_info(struct rga_job *job,
+				   struct rga_job_task_buffers *buffers,
+				   struct rga_req *req)
 {
 	ktime_t timestamp = ktime_get();
 
 	if (job->flags & RGA_JOB_DEBUG_FAKE_BUFFER) {
-		rga_mm_free_fake_buffer(job);
+		rga_mm_free_fake_buffer(job, buffers);
 
 		if (DEBUGGER_EN(TIME))
 			rga_job_log(job, "free fake buffer cost %lld us\n",
@@ -2363,13 +2369,13 @@ void rga_mm_unmap_job_info(struct rga_job *job)
 	}
 
 	if (job->flags & RGA_JOB_USE_HANDLE) {
-		rga_mm_put_handle_info(job);
+		rga_mm_put_handle_info(job, buffers, req);
 
 		if (DEBUGGER_EN(TIME))
 			rga_job_log(job, "put buffer_handle info cost %lld us\n",
 				ktime_us_delta(ktime_get(), timestamp));
 	} else {
-		rga_mm_unmap_buffer_info(job);
+		rga_mm_unmap_buffer_info(job, buffers);
 
 		if (DEBUGGER_EN(TIME))
 			rga_job_log(job, "unmap buffer cost %lld us\n",
@@ -2377,6 +2383,51 @@ void rga_mm_unmap_job_info(struct rga_job *job)
 	}
 }
 
+int rga_mm_map_job_info(struct rga_job *job)
+{
+	int ret = 0;
+	int i;
+	struct rga_job_task_buffers *task_buffers;
+	struct rga_req *req;
+
+	for (i = 0; i < job->task_count; i++) {
+		req = &job->task_list[i];
+		task_buffers = &job->task_buffers[i];
+
+		ret = rga_mm_map_task_info(job, task_buffers, req);
+		if (ret < 0) {
+			rga_job_err(job, "task[%d] failed to map job info\n", i);
+			goto err_unmap_job_info;
+		}
+	}
+
+	return 0;
+
+err_unmap_job_info:
+	for (; i > 0; i--) {
+		task_buffers = &job->task_buffers[i - 1];
+		req = &job->task_list[i - 1];
+
+		rga_mm_unmap_task_info(job, task_buffers, req);
+	}
+
+	return ret;
+}
+
+void rga_mm_unmap_job_info(struct rga_job *job)
+{
+	int i;
+	struct rga_job_task_buffers *task_buffers;
+	struct rga_req *req;
+
+	for (i = 0; i < job->task_count; i++) {
+		task_buffers = &job->task_buffers[i];
+		req = &job->task_list[i];
+
+		rga_mm_unmap_task_info(job, task_buffers, req);
+	}
+}
+
 /*
  * rga_mm_import_buffer - Importing external buffer into the RGA driver
  *
diff --git a/drivers/video/rockchip/rga3/rga_policy.c b/drivers/video/rockchip/rga3/rga_policy.c
index 2554f4bf6389..83909898a19f 100644
--- a/drivers/video/rockchip/rga3/rga_policy.c
+++ b/drivers/video/rockchip/rga3/rga_policy.c
@@ -297,23 +297,18 @@ static bool rga_check_rotate(struct rga_job *job, const struct rga_hw_data *data
 	return true;
 }
 
-int rga_job_assign(struct rga_job *job)
+static int rga_task_assign(struct rga_job *job, struct rga_req *rga_base)
 {
-	struct rga_img_info_t *src0 = &job->rga_command_base.src;
-	struct rga_img_info_t *src1 = &job->rga_command_base.pat;
-	struct rga_img_info_t *dst = &job->rga_command_base.dst;
-
-	struct rga_req *rga_base = &job->rga_command_base;
+	struct rga_img_info_t *src0 = &rga_base->src;
+	struct rga_img_info_t *src1 = &rga_base->pat;
+	struct rga_img_info_t *dst = &rga_base->dst;
 	const struct rga_hw_data *data;
 	struct rga_scheduler_t *scheduler = NULL;
 
 	int feature;
-	int core = RGA_NONE_CORE;
 	int optional_cores = RGA_NONE_CORE;
 	int specified_cores = RGA_NONE_CORE;
 	int i;
-	int min_of_job_count = -1;
-	unsigned long flags;
 	int need_swap = false;
 
 	/* assigned by userspace */
@@ -477,6 +472,35 @@ int rga_job_assign(struct rga_job *job)
 		return -1;
 	}
 
+	return optional_cores;
+}
+
+int rga_job_assign(struct rga_job *job)
+{
+	int i;
+	int ret;
+	int core = RGA_NONE_CORE;
+	int optional_cores = RGA_CORE_MASK;
+	int min_of_job_count = -1;
+	unsigned long flags;
+	struct rga_req *user_req;
+	struct rga_scheduler_t *scheduler;
+
+	for (i = 0; i < job->task_count; i++) {
+		user_req = &job->task_list[i];
+
+		ret = rga_task_assign(job, user_req);
+		if (ret < 0) {
+			rga_job_err(job, "failed to assign task %d\n", i);
+			return ret;
+		}
+
+		if (DEBUGGER_EN(MSG))
+			rga_job_log(job, "task[%d] assign optional cores = %#x\n", i, ret);
+
+		optional_cores &= ret;
+	}
+
 	for (i = 0; i < rga_drvdata->num_of_scheduler; i++) {
 		scheduler = rga_drvdata->scheduler[i];
 
